### PR TITLE
Add tree factory method for build from flat source

### LINKED ISSUE #2122: Method for build Tree from flat source
Building tree from flat source is common task, as I think: items from database query and so on.
But currently building tree from this type of source is complicated.
I suppose to create something like `Tree#ofAll(Iterable<T> source, Function<T, ID> idMapper, Function<T, ID> parentMapper)` (`super`/`extends` hidden for simplicity now).
I already has implementation based on #2121 and can add PR.

### PR DESCRIPTION
**Pull request by valery1707**
Fix #2122

### CODE REVIEW

**Comment by valery1707** [comment]
Partial implementation: at least Javadoc is incomplete

File: vavr/src/main/java/io/vavr/collection/Tree.java [hunk 1]
```diff
@@ -228,11 +228,42 @@
...
         return Tree.of(seed, Stream.of(seed).flatMap(descend).map(children -> recurse(children, descend)));
     }
 
+    /**
+     * Build a {@code List} with roots of {@code Tree} from flat source.
+     * <p>
+     * todo Example:
+     * <pre>{@code
+     *
+     * }</pre>
```
Comment (danieldietrich) [inline vavr/src/main/java/io/vavr/collection/Tree.java]: Please add the following example:

```java
// = [(1, null, "I"), (2, 1, "II"), (3, 1, "III"), (4, 2, "IV"), (5, 2, "V")]
List<MenuItem> items = ...; // MenuItem(id, parentId, label)

//      I
//     / \
//   II  III
//   /\
//  IV V
Tree<MenuItem> menu = Tree.build(items, MenuItem::getId, MenuItem::getParentId);
```

**Comment by danieldietrich** [comment]
@valery1707 ping...

**Comment by valery1707** [comment]
@danieldietrich sorry for long delay :(
I update Javadoc and add to `Value` method `toTree(Function, Function)` for simplify converting `Value` to `tree` with `idMapper` and `parentMapper`.

This changes made in separate commit for simplify review, if all is good to you I will combine than together for beautiful change history

**Comment by valery1707** [comment]
JDK9 build is return strange error:
> error: error while loading package, Missing dependency 'object java.lang.Object in compiler mirror', required by /home/travis/.m2/repository/org/scala-lang/scala-library/2.10.5/scala-library-2.10.5.jar(scala/runtime/package.class)
> error: scala.reflect.internal.MissingRequirementError: object java.lang.Object in compiler mirror not found.

**Comment by valery1707** [comment]
JDK8 build I will fix soon

**Comment by valery1707** [comment]
@danieldietrich build for JDK8 fixed, please review changes

**Comment by danieldietrich** [comment]
Hi @valery1707, thank you for applying the changes! We don‘t have any time pressure, so I don‘t mind about delays. In fact I‘m also delaying things 🙈
The history is clean, I always make squash pulls instead of merge pulls.
Thx!

**Pull request merged**